addons/stock/models/stock_move_line.py

```diff
@@ -586,31 +587,35 @@ def _action_done(self):
 
         # Now, we can actually move the quant.
         ml_ids_to_ignore = OrderedSet()
-        for ml in mls_todo:
-            if ml.product_id.type == 'product':
-                rounding = ml.product_uom_id.rounding
-
-                # if this move line is force assigned, unreserve elsewhere if needed
-                if not ml.move_id._should_bypass_reservation(ml.location_id) and float_compare(ml.qty_done, ml.reserved_uom_qty, precision_rounding=rounding) > 0:
-                    qty_done_product_uom = ml.product_uom_id._compute_quantity(ml.qty_done, ml.product_id.uom_id, rounding_method='HALF-UP')
-                    extra_qty = qty_done_product_uom - ml.reserved_qty
-                    ml._free_reservation(ml.product_id, ml.location_id, extra_qty, lot_id=ml.lot_id, package_id=ml.package_id, owner_id=ml.owner_id, ml_ids_to_ignore=ml_ids_to_ignore)
+        key = itemgetter('product_id', 'location_id', 'location_dest_id', 'lot_id', 'package_id', 'result_package_id', 'owner_id')
+        for (product_id, location_id, location_dest_id, lot_id, package_id, result_package_id, owner_id), move_lines_list in groupby(mls_todo, key=key):
+            mls = self.env['stock.move.line'].concat(*move_lines_list)
+            if product_id.type == 'product':
+                for ml in mls:
+                    rounding = ml.product_uom_id.rounding
+                    # if this move line is force assigned, unreserve elsewhere if needed
+                    if not ml.move_id._should_bypass_reservation(location_id) and float_compare(ml.qty_done, ml.reserved_uom_qty, precision_rounding=rounding) > 0:
```

Comment:
you can move the `should_bypass_reservation` outise the loop on `mls` as it depends only on the location and product